src/utils/ddb/ddb_vos.c

```diff
@@ -2040,3 +2043,411 @@ dv_dev_replace(const char *db_path, uuid_t old_devid, uuid_t new_devid)
+				DP_RC(rc));
+			goto out;
+		}
+		rc = fallocate(fd, 0, 0, rdb_blob_sz);
+		close(fd);
+		if (rc) {
+			rc = daos_errno2der(errno);
+			D_ERROR("fallocate on rdb file %s failed:" DF_RC "\n", rdb_path, DP_RC(rc));
+			goto out;
+		}
+		D_FREE(rdb_path);
+	}
+
+	/** move away from NEWBORNS dir */
+	rc = rename(pool_newborn_path, pool_path);
+	if (rc < 0) {
+		rc = daos_errno2der(errno);
+		D_ERROR(DF_UUID ": failed to rename pool directory: " DF_RC "\n",
+			DP_UUID(pool_uuid), DP_RC(rc));
+		goto out;
+	}
+
+	/** make sure the rename is persistent */
+	// (void)dir_fsync(pool_path);
+
+out:
+	D_FREE(pool_newborn_path);
+	D_FREE(pool_path);
+
+	return rc;
+}
+
+static int
+ddb_tmpfs_setup(const char *path)
+{
+	struct smd_pool_info *pool_info = NULL;
+	struct smd_pool_info *tmp;
+	d_list_t              pool_list;
+	int                   rc = 0;
+	int                   pool_list_cnt;
+	daos_size_t           rdb_blob_sz = 0;
+	mode_t                stored_mode;
+
+	D_ASSERT(bio_nvme_configured(SMD_DEV_TYPE_META));
+	/** create the path string */
+	D_ASPRINTF(newborns_path, "%s/NEWBORNS", path);
+	if (newborns_path == NULL)
+		D_GOTO(err, rc = -DER_NOMEM);
+	D_ASPRINTF(zombies_path, "%s/ZOMBIES", path);
+	if (zombies_path == NULL)
+		D_GOTO(err_newborns, rc = -DER_NOMEM);
+
+	stored_mode = umask(0);
+	/** create NEWBORNS directory if it does not exist already */
+	rc = mkdir(newborns_path, S_IRWXU);
+	if (rc < 0 && errno != EEXIST) {
+		D_ERROR("failed to create NEWBORNS dir: %d\n", errno);
+		umask(stored_mode);
+		D_GOTO(err_zombies, rc = daos_errno2der(errno));
+	}
+
+	/** create ZOMBIES directory if it does not exist already */
+	rc = mkdir(zombies_path, S_IRWXU);
+	if (rc < 0 && errno != EEXIST) {
+		D_ERROR("failed to create ZOMBIES dir: %d\n", errno);
+		umask(stored_mode);
+		D_GOTO(err_zombies, rc = daos_errno2der(errno));
+	}
+	umask(stored_mode);
+
+	D_INIT_LIST_HEAD(&pool_list);
+	rc = smd_pool_list(&pool_list, &pool_list_cnt);
+	if (rc != 0) {
+		D_ERROR("Failed to get pool info list from SMD\n");
+		return rc;
+	}
+
+	d_list_for_each_entry_safe(pool_info, tmp, &pool_list, spi_link) {
+		D_INFO("recreating vos files for pool " DF_UUID "\n", DP_UUID(pool_info->spi_id));
+		rc = smd_rdb_get_blob_sz(pool_info->spi_id, &rdb_blob_sz);
+		if (rc && (rc != -DER_NONEXIST)) {
+			D_ERROR(DF_UUID ": failed to extract the size of rdb file: " DF_RC "\n",
+				DP_UUID(pool_info->spi_id), DP_RC(rc));
+			goto out;
+		}
+
+		D_ASSERT(pool_info->spi_scm_sz > 0);
+		rc = tgt_recreate(path, pool_info->spi_id, pool_info->spi_scm_sz,
+				  pool_info->spi_tgt_cnt[SMD_DEV_TYPE_META], rdb_blob_sz);
+		if (rc)
+			goto out;
+	}
+out:
+	d_list_for_each_entry_safe(pool_info, tmp, &pool_list, spi_link) {
+		d_list_del(&pool_info->spi_link);
+		/* Frees spi_tgts, spi_blobs, and pool_info */
+		smd_pool_free_info(pool_info);
+	}
+
+err_zombies:
+	D_FREE(zombies_path);
+err_newborns:
+	D_FREE(newborns_path);
+err:
+	return rc;
+}
```

Comment:
I was thinking that "setup" will be called for each VOS file since ddb can only open VOS pool in standalone mode.

Now it looks to me that "setup" only needs be called once to setup all VOS files, that sounds reasonable to me (user don't have to call setup on each vos file, also user will be able to list all existing files when setup done).

But now I tend to think we'd spend some effort on removing all above duplicated code. :)